libs/@local/hash-graph-types/graph_types/data_type.py

```diff
@@ -49,15 +46,15 @@ async def create_model(
             return cached
 
         schema = await graph.get_data_type(self.ref, actor_id=actor_id)
+        model = await schema.create_data_type(actor_id=actor_id, graph=graph)
+
+        # Potential race condition here, another coroutine could have created the model
+        # before we did on the previous await statement.
+        # Just check if we have it in the cache again, this way we won't
+        # accidentally create the same model twice.
+        if cached := self.cache.get(self.ref):
+            return cached
```

Comment:
I tried this before and this does not work as expected. For example for the entity type `https://blockprotocol.org/@examples/types/entity-type/person/v/1` this just generated this schema:
```json
{
  "$defs": {
    "https_blockprotocol_org_blockprotocol_types_property_type_name_v_1": {
      "$id": "https://blockprotocol.org/@blockprotocol/types/property-type/name/v/1",
      "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
      "allOf": [
        {
          "$ref": "#/$defs/pydantic__main__https_blockprotocol_org_blockprotocol_types_data_type_text_v_1_4578735440"
        }
      ],
      "description": "A word or set of words by which something is known, addressed, or referred to.",
      "kind": "propertyType",
      "title": "Name"
    },
    "https_blockprotocol_org_examples_types_property_type_e_mail_v_1": {
      "$id": "https://blockprotocol.org/@examples/types/property-type/e-mail/v/1",
      "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
      "allOf": [
        {
          "$ref": "#/$defs/pydantic__main__https_blockprotocol_org_blockprotocol_types_data_type_text_v_1_4572375792"
        }
      ],
      "description": "An e-mail address.",
      "kind": "propertyType",
      "title": "E-Mail"
    },
    "pydantic__main__https_blockprotocol_org_blockprotocol_types_data_type_text_v_1_4572375792": {
      "$id": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1",
      "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      "description": "An ordered sequence of characters",
      "kind": "dataType",
      "title": "Text",
      "type": "string"
    },
    "pydantic__main__https_blockprotocol_org_blockprotocol_types_data_type_text_v_1_4578735440": {
      "$id": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1",
      "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
      "description": "An ordered sequence of characters",
      "kind": "dataType",
      "title": "Text",
      "type": "string"
    }
  },
  "$id": "https://blockprotocol.org/@examples/types/entity-type/person/v/1",
  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
  "additionalProperties": false,
  "description": "An extremely simplified representation of a person or human being.",
  "kind": "entityType",
  "properties": {
    "https://blockprotocol.org/@examples/types/property-type/e-mail/": {
      "$ref": "#/$defs/https_blockprotocol_org_examples_types_property_type_e_mail_v_1"
    },
    "https://blockprotocol.org/@blockprotocol/types/property-type/name/": {
      "$ref": "#/$defs/https_blockprotocol_org_blockprotocol_types_property_type_name_v_1"
    }
  },
  "required": [
    "https://blockprotocol.org/@blockprotocol/types/property-type/name/"
  ],
  "title": "Person",
  "type": "object"
}
```

The `text` data type appears twice here.

Reply:
I unified the cache logic in [9d30d65](https://github.com/hashintel/hash/pull/2776/commits/9d30d651ed2a11e9333d3d57913227bf72d764e8), could you try again?

Reply:
This is probably worth moving into a separated PR.

Reply:
This works for me (both did) but I think the issue is that the temporal implementation may use threads and asyncio primitives are not thread safe™

Reply:
But I think this could be the case. Even though it's likely not possible to do the things in parallel, it may be done concurrently.

Reply:
I mean, this is why I tried to use the synchronization primitives for `Event`, but getting this right in Python is a lot harder than say... Rust.

Reply:
> it may be done concurrently.

To guard against that exact thing no await point is before checking, so that shouldn't happen in a non-threaded environment. Anyway. More investigation is needed 🔍